src/cmdline/mod.rs

```diff
@@ -234,24 +221,110 @@ impl Cmdline {
     /// ```rust
     /// # use linux_loader::cmdline::*;
     /// # use std::ffi::CString;
-    /// let mut cl = Cmdline::new(100);
-    /// cl.insert_str("foobar");
-    /// let cl_cstring = CString::new(cl).unwrap();
-    /// assert_eq!(cl_cstring.to_str().unwrap(), "foobar");
+    /// let mut cl = Cmdline::new(100).unwrap();
+    /// cl.insert_str("foobar").unwrap();
+    /// assert_eq!(cl.as_cstring().unwrap().as_bytes_with_nul(), b"foobar\0");
     /// ```
     pub fn insert_str<T: AsRef<str>>(&mut self, slug: T) -> Result<()> {
-        let s = slug.as_ref();
+        // Step 1: Check if the string provided is a valid boot arg string and remove any
```

Comment:
These two `insert`s are really similar so I think you can comfortably extract the common bits into (yet another) helper - the flow is basically:
1. check that `slug` fits
2. put `slug` in either `boot_args` or `init_args`

So you could compress step 1. into a checker and just pass it how many extra chars it needs to accommodate, something like

```rust
fn check_size<T: AsRef<str>>(&mut self, slug: &T, num_extra_chars: usize) -> Result<()> {
        // Step 1: Check if the string provided is a valid boot arg string and remove any
        // leading or trailing whitespaces.
        let s = slug.as_ref().trim();
        valid_str(s)?;

        // Step 2: Check if cmdline capacity is not exceeded when inserting the arg
        // string provided.
        let cmdline_size = self.get_null_terminated_representation_size()
        .checked_add(num_extra_chars).ok_or(Error::TooLarge)?
        .checked_add(s.len()).ok_or(Error::TooLarge)?;

        if cmdline_size > self.capacity {
            Err(Error::TooLarge)
        } else {
            Ok(())
        }
    }
```

Then call it like this -

```rust
pub fn insert_str<T: AsRef<str>>(&mut self, slug: T) -> Result<()> {
        self.check_size(&slug, if self.boot_args.is_empty() { 0 } else { 1 })?;

        if !self.boot_args.is_empty() {
        // etc
    }

pub fn insert_init_args<T: AsRef<str>>(&mut self, slug: T) -> Result<()> {
        self.check_size(&slug, if self.init_args.is_empty() {
            INIT_ARGS_SEPARATOR.len()
        } else {
            1
        })?;

        if !self.init_args.is_empty() {
        // etc
}
````